### PR TITLE
Track features from the backtrace of exceptions

### PR DESCRIPTION
**Pull request by metcalf**
When the action run by Zeus raises an exception, we report `all_features` and crash. Since files do not get added to Ruby's `$LOADED_FEATURES` list unless the are required without error, this means that Zeus may not reload when you fix the underlying issue. This is mostly plastered over by the fact that Zeus accumulates features across all runs of a command so if your code loads once, it'll know what to reload on if it crashes the next time. This doesn't work on first boot-up and constrains some possible improvements to file tracking in the SlaveNode. At Stripe, we're currently monkey patching `Kernel#require_relative` in our `Plan` to work around this issue.

This changes `features_loaded_by` to catch exceptions in the executing block and add the lines of the backtrace into the list of features reported to Zeus. This should provide an accurate picture of the files that were actually accessed leading to the crash. The existing shared test helpers were stubbing out `Kernel#load`, breaking some `LoadTracking` tests. I've moved spec helpers to be scoped to the modules in which they're used and killed the shared file.

### CODE REVIEW

**Comment by metcalf** [comment]
@ptarjan, up for reviewing this? You figured out our `require` monkey patch in the first place.

File: rubygem/lib/zeus/load_tracking.rb [hunk 1]
```diff
@@ -2,10 +2,32 @@ module Zeus
...
+        if err && err.backtrace
+          err_features += err.backtrace.map { |b| b.split(':').first }
+                             .select { |f| f.start_with?('/') }
+                             .take_while { |f| f != __FILE__ }
+        end
+
+        new_features = all_features + err_features - old_features
+        new_features.uniq!
+
+        [new_features, err]
```
Comment (ptarjan) [inline rubygem/lib/zeus/load_tracking.rb]: I'm not too in love with this and usually prefer a dictionary, but your call

File: rubygem/spec/load_tracking_spec.rb [hunk 2]
```diff
@@ -30,12 +32,74 @@
...
+        new_files, = Zeus::LoadTracking.features_loaded_by do
           $untracked_features << "an_untracked_feature.rb"
         end
 
         expect(new_files).to eq(["an_untracked_feature.rb"])
       end
     end
   end
+
+  describe '.features_loaded_by' do
```
Comment (ptarjan) [inline rubygem/spec/load_tracking_spec.rb]: happydance

File: rubygem/spec/m_spec.rb [hunk 3]
```diff
@@ -1,11 +1,36 @@
...
+      failure_message_when_negated do |_block|
+        "expected block not to call exit(#{exp_code})"
+      end
+      description do
+        "expect block to call exit(#{exp_code})"
+      end
+    end
+
     before(:each) do
+      allow(Dir).to receive(:glob).and_return(["path/to/file.rb"])
```
Comment (ptarjan) [inline rubygem/spec/m_spec.rb]: Should these only be in the test places they are used?

**Comment by ptarjan** [comment]
LGTM

File: rubygem/spec/load_tracking_spec.rb [hunk 4]
```diff
@@ -30,12 +32,74 @@
...
+      end
+
+      it 'tracks loads' do
+        expect_to_load([expand_asset_path('load.rb')]) do
+          load expand_asset_path('load.rb')
+        end
+      end
+    end
+
+    context 'loading invalid code' do
```
Comment (latortuga) [inline rubygem/spec/load_tracking_spec.rb]: :+1: :+1: :100: :heart: :yellow_heart: :blue_heart: this is going to be an AWESOME addition to zeus

File: rubygem/lib/zeus/load_tracking.rb (hunk 1 above)
Comment (metcalf) [inline rubygem/lib/zeus/load_tracking.rb]: `Struct.new(:features, :error)` maybe?

Comment (metcalf) [inline rubygem/lib/zeus/load_tracking.rb]: Thought about this a bit more, I'm reasonably indifferent but in this case the particular return pattern is consistent with what all the Go in this codebase is doing (value, err). Given the relatively light preferences either way, I'm going to stick with it.

**Pull request merged**